### PR TITLE
Regexp: Transpile `\d`, `\w` to Java's definitions

### PR DESCRIPTION
**Pull request by andygrove**
Closes https://github.com/NVIDIA/spark-rapids/issues/5530

cuDF and Java have different interpretations for `\d`, `\w` so we now transpile them to Java's definition to ensure that we get the expected results.

### CODE REVIEW

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
@anthony-chang could you review?

**Review by anthony-chang — APPROVED** [review]

**Comment by andygrove** [comment]
build

**Review by revans2 — COMMENTED** [review]
Also do we want to add some explicit tests for this in python that are not fuzzing tests?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 1]
```diff
@@ -792,6 +792,16 @@ class CudfRegexTranspiler(mode: RegexMode) {
         }
 
       case RegexEscaped(ch) => ch match {
+        case 'd' =>
+          // cuDF is not compatible with Java for \d  so we transpile to Java's definition
+          // https://github.com/rapidsai/cudf/issues/10894
+          RegexCharacterClass(negated = false, ListBuffer(RegexCharacterRange('0', '9')))
+        case 'w' =>
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:799]: In a follow on PR?

File: integration_tests/src/main/python/string_test.py [hunk 2]
```diff
@@ -939,6 +939,26 @@ def test_regexp_octal_digits():
             ),
         conf=_regexp_conf)
 
+def test_regexp_replace_digit():
+    gen = mk_str_gen('[a-z]{0,2}[0-9]{0,2}') \
+        .with_special_case('䤫畍킱곂⬡❽ࢅ獰᳌蛫青')
```
Comment (andygrove) [inline integration_tests/src/main/python/string_test.py:944]: @revans2 is it ok to add Unicode test data like this or should I convert it to `\unnnn` syntax?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 3]
```diff
@@ -792,6 +792,16 @@ class CudfRegexTranspiler(mode: RegexMode) {
...
+          // cuDF is not compatible with Java for \w so we transpile to Java's definition
+          RegexCharacterClass(negated = false, ListBuffer(
+            RegexCharacterRange('a', 'z'),
+            RegexCharacterRange('A', 'Z'),
+            RegexCharacterRange('0', '9')))
```
[the remark this answers is not in the document]
Comment (andygrove) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:807]: I updated this

File: integration_tests/src/main/python/string_test.py (hunk 2 above)
Comment (revans2) [reply integration_tests/src/main/python/string_test.py:944]: I think it is fine.

**Review by revans2 — DISMISSED** [review]

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
Test fails with the following information. I do not see this locally so either the test is somehow not deterministic or there is some environment difference in CI around Unicode but I am really just guessing at this point. 

```
[2022-05-19T18:00:53.673Z] - AST fuzz test - regexp_replace - full unicode input *** FAILED ***
[2022-05-19T18:00:53.673Z]   javaPattern=\z\u20282, cudfPattern=$\u20282, input='????', cpu=????, gpu=???????????? (RegularExpressionTranspilerSuite.scala:733)
```

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
@revans2 I had to disable the scala fuzz tests because they fail in CI (but not locally) and I don't know why yet. I filed https://github.com/NVIDIA/spark-rapids/issues/5549 to track this. I think we are ok to merge this PR without these tests though since there are integration tests covering the changes.

**Review by revans2 — APPROVED** [review]
I am okay with disabling the test for now, but we should treat it as a P1 until we understand exactly where it is failing and why.

**Comment by andygrove** [comment]
I am moving the new scala unicode fuzz tests to a separate PR (will be up in the next hour) and I marked the issue https://github.com/NVIDIA/spark-rapids/issues/5549 as a P1.

This PR now supports `\W` and `\D` as well.

**Review by revans2 — DISMISSED** [review]

**Comment by andygrove** [comment]
I spoke too soon `\D` and `\W` .. will create a separate PR for that

**Comment by andygrove** [comment]
build

**Review by revans2 — APPROVED** [review]

**Pull request merged**